cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -173,28 +173,94 @@ struct MapRandomAccessWrapper {
 inline std::shared_ptr<std::unordered_map<StreamId, AtomKey>> batch_get_specific_version(
     const std::shared_ptr<Store>& store,
     const std::shared_ptr<VersionMap>& version_map,
     const std::map<StreamId, VersionId>& sym_versions,
-    bool include_deleted = true) {
+    bool include_deleted = true,
+    bool include_referenced_in_snapshot_only = false) {
     ARCTICDB_SAMPLE(BatchGetLatestVersion, 0)
     auto output = std::make_shared<std::unordered_map<StreamId, AtomKey>>();
-    auto mutex = std::make_shared<std::mutex>();
+    auto output_mutex = std::make_shared<std::mutex>();
+    auto tombstoned_vers = std::make_shared<std::vector<std::pair<StreamId, AtomKey>>>(); // symbol, version key
+    auto tombstoned_vers_mutex = std::make_shared<std::mutex>();
 
     MapRandomAccessWrapper wrapper{sym_versions};
     submit_tasks_for_range(wrapper, [store, version_map](auto& sym_version) {
             LoadParameter load_param{LoadType::LOAD_DOWNTO, static_cast<SignedVersionId>(sym_version.second)};
             return async::submit_io_task(CheckReloadTask{store, version_map, sym_version.first, load_param});
         },
-        [output, include_deleted, mutex](auto sym_version, const std::shared_ptr<VersionMapEntry>& entry) {
-        auto index_key = find_index_key_for_version_id(sym_version.second, entry, include_deleted);
-        if (index_key) {
-            std::lock_guard lock{*mutex};
-            (*output)[sym_version.first] = *index_key;
-        }
-    });
+        [output, include_deleted, include_referenced_in_snapshot_only, output_mutex, store, tombstoned_vers, tombstoned_vers_mutex]
+                    (auto sym_version, const std::shared_ptr<VersionMapEntry>& entry) {
+            auto key_and_is_tombstoned = find_index_key_for_version_id_and_tombstone_status(sym_version.second, entry);
+            if (key_and_is_tombstoned) {
+                auto [key, is_tombstoned] = key_and_is_tombstoned.value();
+                bool is_valid_key = include_deleted || !is_tombstoned;
+                if (is_valid_key) {
+                    std::lock_guard lock{*output_mutex};
+                    (*output)[sym_version.first] = key;
+                }
+                else if (include_referenced_in_snapshot_only && is_tombstoned) { // Need to allow tombstoned version but referenced in other snapshot(s) can be "re-snapshot"
+                    log::version().warn("Version {} for symbol {} is tombstoned, need to check snapshots (this can be slow)", sym_version.second, sym_version.first);
+                    std::lock_guard lock{*tombstoned_vers_mutex};
+                    tombstoned_vers->emplace_back(sym_version.first, key);
+                }
+            }
+        });
+
+    if (!tombstoned_vers->empty()) {
+        const auto snap_map = get_master_snapshots_map(store);
+        submit_cpu_task_for_range(*tombstoned_vers, [&snap_map, &sym_versions, output, output_mutex](const auto& tombstoned_ver) {
```

Comment:
Does this really need to be parallelised?

Reply:
Not really as I have mentioned in the PR description. Gonna it back